🔧 MAINTAIN: Move to flit for PEP 621 package builds

choldgraf opened the pull request:
This removes our `setup.py` file, which we had marked for removal once https://github.com/pypa/packaging-problems/issues/256 was resolved (and it is now resolved!)

I also updated a couple of action jobs since we needed to update our build distribution action.

I tested `pip install -e .` locally without `setup.py` and it seemed to work as expected. Will leave this open for a few days in case others think this is not a good action to take yet.

chrisjsewell commented:
I would also just move to [flit](https://flit.readthedocs.io) and remove setup.cfg and MANIFEST.in

choldgraf commented:
@chrisjsewell have you found any nice tutorials on that? I'm happy to give it a try but it'd be a first for me.

chrisjsewell commented:
Well just the flit documentation should be mainly all you need, you can also try https://ini2toml.readthedocs.io/en/latest/readme.html, and I've already done this in e.g. https://github.com/aiidateam/aiida-core/pull/5312

chrisjsewell commented:
the `MANIFEST.in` can basically just be deleted, since flit handles all that automatically
Plus this could be removed: https://github.com/executablebooks/jupyter-book/blob/abf118388834b2c97ec74aac4fae095e35481a19/.pre-commit-config.yaml#L22-L34

This would also need updating, but wouldn't be too hard: https://github.com/executablebooks/jupyter-book/blob/abf118388834b2c97ec74aac4fae095e35481a19/.pre-commit-config.yaml#L36-L49

choldgraf commented:
will try to give it a shot tonight and ping you if anything comes up

choldgraf commented on pyproject.toml:
```diff
@@ -1,6 +1,119 @@
+    "docutils>=0.15,<0.18",
+    "jsonschema<4",
+    # Include Jupytext to ensure users have the right version, even if not strictly necessary
+    "jupytext>=1.11.2,<1.12", # markdown-it-py~=1 required and support from 1.11.2
```
An aside - I noticed that we're pinning jupytext on a minor version rather than a major version. It's now on `1.13.7` - do we still need to keep this?

chrisjsewell replied:
Well, we kind of don't need the jupytext dependency at all.
The lower pin was added when markdown-it-py was only an extra of jupytext, so you could end up with non-matching versions, and jupytext would just fail to convert properly.
But now, markdown-it-py isa first-class dependency of jupytext: https://github.com/mwouts/jupytext/blob/8cfa66bef380bf6d54479a06b092da30dbc9a7c4/setup.py#L38

choldgraf commented:
OK I believe that I got a working `flit` configuration setup. I was able to successfully run `flit build` and it built without any errors. Looks reasonable to you @chrisjsewell ?

chrisjsewell commented on pyproject.toml:
```diff
@@ -1,6 +1,119 @@
+build-backend = "flit_core.buildapi"
+
+[project]
+name = "jupyter_book"
```
The only "concern" of this name change, from `jupyter-book` to `jupyter_book`, is that although pip treats them the same (normalising `-` and `_`), conda for example does not. So I'm not sure if there would definitely be any issue.

You can keep the name as `jupyter-book`, and set the module name separately, with:

```yaml
[tool.flit.module]
name = "jupyter_book"
```

choldgraf replied:
Ah perfect, will do this.

chrisjsewell commented:
Apart from two comments, looks good 👍

choldgraf commented:
OK I addressed both comments w/ the changes suggested. Jupytext is now in our `[sphinx]` dependencies, since it's needed for building from an Rmd file. Also added a note for others to install it if they want to use RMD
